File: HackAssembler/Assembler.cs
using System;
using System.Collections.Generic;

namespace HackAssembler
{
    public class Assembler
    {
        private Parser parser;

        //private SymbolTable symbolTable;

        public List<string> HackCode { get; private set; }

        public Assembler(string sourcePath)
        {
            HackCode = new List<string>();
            parser = new Parser(sourcePath);
            //symbolTable = new SymbolTable();
        }

        public void Run()
        {
            HackCode.Clear();

            while (parser.HasMoreCommands)
            {
                parser.Advance();

                switch (parser.CurrentCommandType)
                {
                    case Parser.CommandType.C_COMMAND:
                        //               +--------comp-------+ +-dest--+ +-jump-+
                        // Binary: 1 1 1 a  c1 c2 c3 c4  c5 c6 d1 d2  d3 j1 j2 j3
                        var cmd = "111";
                        cmd += Code.BitArrayToString(Code.Comp(parser.Comp));
                        cmd += Code.BitArrayToString(Code.Dest(parser.Dest));
                        cmd += Code.BitArrayToString(Code.Jump(parser.Jump));
                        HackCode.Add(cmd);
                        break;

                    case Parser.CommandType.A_COMMAND:
                        //           +----value (v = 0 or 1)-----+
                        // Binary: 0 v v v v v v v v v v v v v v v
                        int address = Convert.ToInt16(parser.Symbol);
                        HackCode.Add("0" + Convert.ToString(address, 2).PadLeft(15, '0'));
                        break;
                }
            }
        }

        public bool Save(string binaryFile)
        {
            if (HackCode.Count == 0)
                return false;
            System.IO.File.WriteAllLines(binaryFile, HackCode);
            return true;
        }
    }
}

File: cpp/assembler.cpp
#include <QTextStream>
#include "assembler.h"
#include "code.h"

bool Assembler::open(const char* sourceFile)
{
    if (!m_parser.openSourceFile(sourceFile))
        return false;

    m_hackCode.clear();
    m_symbolTable.clear();
    m_errorMessage.clear();

    return true;
}

bool Assembler::save(const char* binaryFile)
{
    QFile outputFile(binaryFile);
    if (!outputFile.open(QIODevice::WriteOnly | QIODevice::Text))
        return false;

    QTextStream out(&outputFile);
    for (const QByteArray& line : m_hackCode)
        out << line << "\n";

    return true;
}

void Assembler::run()
{
    processLabels();

    while (m_parser.hasMoreCommands()) {
        m_parser.advance();
        uint address;

        switch (m_parser.commandType()) {
        case Parser::C_COMMAND:
            m_hackCode.append("111" +
                              Code::comp(m_parser.comp()) +
                              Code::dest(m_parser.dest()) +
                              Code::jump(m_parser.jump()));
            break;

        case Parser::A_COMMAND:
            bool isNumeric;
            address = m_parser.symbol().toUInt(&isNumeric);
            if (!isNumeric)
                address = m_symbolTable.getAddressWithAddEntry(m_parser.symbol());
            m_hackCode.append(QByteArray::number(address, 2).rightJustified(16, '0'));

        default:
            break;
        }
    }
}

const QList<QByteArray>& Assembler::hackCode() const
{
    return m_hackCode;
}

bool Assembler::hasError() const
{
    return !m_errorMessage.isEmpty();
}

const QByteArray& Assembler::errorMessage() const
{
    return m_errorMessage;
}

void Assembler::processLabels()
{
    uint memoryAddress = 0;

    while (m_parser.hasMoreCommands()) {
        m_parser.advance();

        switch (m_parser.commandType()) {
        case Parser::A_COMMAND:
        case Parser::C_COMMAND:
            memoryAddress++;
            break;

        case Parser::L_COMMAND:
            m_symbolTable.addEntry(m_parser.symbol(), memoryAddress);
            break;

        default:
            break;
        }
    }
    m_parser.reset();
}
